daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/midBitsToEndTransformer.scala

```diff
@@ -0,0 +1,339 @@
+import org.apache.daffodil.processors.LayerCharsetEv
+import org.apache.daffodil.schema.annotation.props.gen.LayerLengthKind
+import org.apache.daffodil.processors.LayerLengthInBytesEv
+import org.apache.daffodil.schema.annotation.props.gen.LayerLengthUnits
+import org.apache.daffodil.processors.LayerBoundaryMarkEv
+import org.apache.daffodil.processors.LayerTransformArgsEv
+import java.io.InputStream
+import java.io.OutputStream
+import org.apache.daffodil.processors.unparsers.UState
+import org.apache.daffodil.processors.parsers.PState
+import org.apache.daffodil.exceptions.Assert
+import org.apache.daffodil.processors.ParseOrUnparseState
+import org.apache.daffodil.processors.parsers.Parser
+
+import org.apache.daffodil.processors.ExplicitLengthEv
+import org.apache.daffodil.io.ExplicitLengthLimitingStream
+import java.io.ByteArrayOutputStream
+import java.io.ByteArrayInputStream
+import org.apache.daffodil.util.MaybeULong
+import org.apache.daffodil.processors.Failure
+
+class midBitsToEndTransformer(
+  layerLengthInBytesEv: LayerLengthInBytesEv,
+  layerTransformArgsEv: LayerTransformArgsEv)
+  extends LayerTransformer {
+  /*
+   * For unparsing, it is more or less unavoidable that we cannot stream, as we need to reach
+   * the end of the input before we can provide the mid-bits that were moved.
+   * In practice, it is not expected that this transform will be used on "large" blocks,
+   * so it is not worth the effort to figure out how to implement some form of streaming.
+   */
+
+  var length: Int = -1
+  var firstBitIndex: Int = -1
+  var numBits: Int = -1
+
+  /*
+   * Note the extra padding bytes. One of these bytes is just to accomodate rounding from integer division.
+   * The other one is because, to simplify the code, a portion of the encoder may attempt to read up to 7 extra bits,
+   * then discard them. (This would happen when the last bit that gets moved is the first bit of a byte in the encoded form;
+   * the encoder will attempt to read enough bits to be able to fully replace the entire byte, then mask out the lower 7 bits)
+   */
+  def tmpByteCount = numBits / 8 + 2
+  def buffSize = length + tmpByteCount
+
+  def byteShift = numBits / 8
+  def bitShift = numBits % 8
+
+  def firstBitByteIndex = (firstBitIndex / 8)
+  def lastBitByteIndex = ((firstBitIndex + numBits) / 8)
+  def firstBitBitIndex = firstBitIndex % 8
+  def lastBitBitIndex = (firstBitIndex + numBits) % 8
+
+  private def buffToHex(buff: Array[Byte]): String = {
+    buff.map(b => String.format("%02x", Byte.box(b))).mkString(" ")
+  }
+
+  private def initialize(state: ParseOrUnparseState) {
+    length = layerLengthInBytesEv.evaluate(state).toInt
+    val argsString = layerTransformArgsEv.evaluate(state)
+      .trim()
+      .split("\\ +")
+    if (argsString.length != 2) {
+      state.SDE(s"""midBitsToEnd layer transform requires exactly 2 arguements, recieved ${argsString.length}: "${argsString.mkString(",")}" """)
```

Comment:
SDE should not be given a string that has things substituted into it like this.

SDE takes a format string and then varargs, for a reason, which is Internationalization. The format string is used as a hash key into internationalized translations of the string, and then the args are selected from the arglist and embedded into various languages in different orders generally.

So don't feed SDE a string in engilish that has the args substituted in. To do so subverts the design for I18n.